pkg/controllers/user/rbac/legacy_grb_cleaner.go

```diff
@@ -19,19 +20,38 @@ func newLegacyGRBCleaner(m *manager) *grbCleaner {
 	return &grbCleaner{m: m}
 }
 
+// TODO  Remove this once no longer needed
+
+// this function addresses issues with grb not being cleaned up that was an issue from v2.2.3 - v2.2.8
+// it works on previously created objects outside of the timeline of normal sync handlers to correct issues with finalizers not be removed on deletion and finalizers not being dropped on cluster deletion
 func (p *grbCleaner) sync(key string, obj *v3.GlobalRoleBinding) (runtime.Object, error) {
 	if key == "" || obj == nil {
 		return nil, nil
 	}
-	if obj.DeletionTimestamp == nil || len(obj.Finalizers) == 0 {
-		return nil, nil
+	if len(obj.Finalizers) == 0 {
+		return obj, nil
 	}
-	if time.Since(obj.DeletionTimestamp.Time) < 1*time.Hour {
-		return nil, nil
-	}
-	obj, err := p.removeStuckFinalizer(obj)
-	if err != nil && !errors.IsNotFound(err) {
-		return nil, err
+	if obj.DeletionTimestamp != nil {
+		if time.Since(obj.DeletionTimestamp.Time) < 1*time.Hour {
+			return obj, nil
+		}
+		obj, err := p.removeStuckFinalizer(obj)
+		if err != nil && !errors.IsNotFound(err) {
+			return obj, err
+		}
+	} else {
+		if time.Since(obj.CreationTimestamp.Time) < 1*time.Hour {
+			return obj, nil
+		}
+
+		clusterList, err := p.m.clusterLister.List("", labels.Everything())
+		if err != nil {
+			return obj, err
+		}
+		obj, err = p.removeFinalizerFromNonExistentCluster(obj, clusterList)
+		if err != nil && !errors.IsNotFound(err) {
+			return obj, err
+		}
```

Comment:
Before leaving, a GRB that has been processed successfully should get an annotation to indicate it's already been cleaned. This annotation should be checked above so we don't process the same one repeatedly.

Reply:
Added annotation